### PR TITLE
Add optin command for Marketo integration

### PR DESCRIPTION
**Pull request by stvstnfrd**

### CODE REVIEW

**Comment by stvstnfrd** [comment]
I added tests.

File: openedx/stanford/djangoapps/auth_lagunita/models.py [hunk 1]
```diff
@@ -3,10 +3,13 @@
 Record extra lagunita-centric user data
 """
 from __future__ import unicode_literals
+import logging
 
 from django.conf import settings
 from django.db import models
+from django.db.models import Q
 
+log = logging.getLogger('auth_lagunita')
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/models.py]: This doesn't appear to be used anywhere in this file?

Comment (stvstnfrd) [reply openedx/stanford/djangoapps/auth_lagunita/models.py]: I must have factored out the logging calls in this file;
I removed this.

File: openedx/stanford/djangoapps/auth_lagunita/api.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+            base_url=self.base_url,
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+        )
+        # pylint: enable=line-too-long
+        log.debug("url: %s", url)
+        access_token = None
+        try:
+            response = requests.get(url)
+        except requests.exceptions.HTTPError as exception:
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/api.py]: Should we catch the base class `requests.exceptions.RequestException` to be more robust? https://stackoverflow.com/a/16511493

File: openedx/stanford/djangoapps/auth_lagunita/api.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+            error = "requests.exceptions: {0}".format(exception)
+        else:
+            log.error(response)
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/api.py]: Why `log.error` rather than `log.info` here?

File: openedx/stanford/djangoapps/auth_lagunita/api.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+        )
+        # pylint: enable=line-too-long
+        log.debug("url: %s", url)
+        access_token = None
+        try:
+            response = requests.get(url)
+        except requests.exceptions.HTTPError as exception:
+            error = "requests.exceptions: {0}".format(exception)
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/api.py]: Should this be logged?

File: openedx/stanford/djangoapps/auth_lagunita/api.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+            log.error("Subscription failed for user=%s", user.email)
+        return subscription_id
+
+    def _create_lead(self, user):
+        """
+        Create a new lead entry, fails if already exists
+        """
+        log.info("Lead creation attempted: %s", user.email)
+        data = {
+            'action': 'createOnly',
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/api.py]: The docs show two endpoints ([first](https://developers.marketo.com/rest-api/lead-database/leads/#create_and_update), [second](https://developers.marketo.com/rest-api/lead-database/leads/#push_lead_to_marketo)) that support adding new leads, but it's not clear to me which one we want here...the code here is hitting the first one, but it doesn't support the `'reason'` and `'source'` params, while second one does, but that one additionally requires a `'programName'` param. Thoughts?

File: openedx/stanford/djangoapps/auth_lagunita/api.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        result, error = self._create_lead(user)
+        if error == 'skipped':
+            log.info("Lead creation exists: %s", user.email)
+            lead_id = self.lookup_lead(user.email)
+        elif error:
+            log.error("Lead creation failed: %s, %s", user.email, error)
+            lead_id = None
+        else:
+            lead_id = result['id']
+            log.error("Lead creation succeeded: %s, %s", user.email, lead_id)
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/api.py]: `log.info` is more appropriate, right?

File: openedx/stanford/djangoapps/auth_lagunita/api.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        log.debug("headers: %s", headers)
+        log.debug("data: %s", data)
+        result = None
+        try:
+            response = request_method(
+                url=url,
+                headers=headers,
+                data=json.dumps(data),
+            )
+        except requests.exceptions.HTTPError as exception:
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/api.py]: same concern as https://github.com/Stanford-Online/edx-platform/pull/903/files#r310767294

File: openedx/stanford/djangoapps/auth_lagunita/api.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+                        error = "API operation failed; {errors}".format(
+                            errors=errors,
+                        )
+                else:
+                    result = data['result']
+                    if len(result):
+                        result = result[0]
+                        if result.get('status') == 'skipped':
+                            result = None
+                            error = 'skipped'
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/api.py]: Do we also want to including the error `reasons`? https://developers.marketo.com/rest-api/error-codes/#record_level_errors

File: openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        self.api_url = kwargs['api_url'] or settings.MARKETO_URL
+        self.client_id = kwargs['client_id'] or settings.MARKETO_CLIENT_ID
+        self.client_secret = kwargs['client_secret'] or settings.MARKETO_CLIENT_SECRET
+        self.dry_run = kwargs['dry_run']
+        self.list_id = kwargs['list_id'] or settings.MARKETO_LIST_ID
+        self.verbosity = kwargs['verbosity']
+
+    def _set_logging_verbosity(self):
+        """
+        Toggle verobsity via CLI
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py]: small typo, `verbosity`

File: openedx/stanford/djangoapps/auth_lagunita/models.py [hunk 10]
```diff
@@ -48,3 +49,24 @@ def __unicode__(self):
...
+    infos = infos.exclude(
+        Q(user__email='') |
+        Q(user__email=None) |
+        Q(user__email__endswith='@example.com') |
+        Q(user__email__endswith='.example.com')
+    )
+    return infos
+
+
+Info.objects.need_subscribed = _need_subscribed
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/models.py]: Django docs only mention creating an entire [separate custom manager](https://docs.djangoproject.com/en/2.2/topics/db/managers/#adding-extra-manager-methods) to handle custom queries, rather than directly modifying the default objects manager. Just curious the reason for the latter approach, is it just a one-off thing?

File: openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            if self.dry_run:
+                log.info("Subscription faked for user=%s", user.email)
+            else:
+                if not api.access_token:
+                    api.refresh_access_token()
+                subscription = api.subscribe(user, self.list_id)
+                if subscription:
+                    info.submitted_marketing_optin = True
+                    info.save()
+            time.sleep(0.5)
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py]: Just checking, this is to adhere with Marketo API rate limits right?

File: openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        if len(need_subscribed):
+            log.info('User lookup succeeded')
+        else:
+            log.info('User lookup empty')
+        return
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py]: Why is this line needed?

File: openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+            else:
+                if not api.access_token:
+                    api.refresh_access_token()
+                subscription = api.subscribe(user, self.list_id)
+                if subscription:
+                    info.submitted_marketing_optin = True
+                    info.save()
+            time.sleep(0.5)
+        if len(need_subscribed):
+            log.info('User lookup succeeded')
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py]: sort of a nit, it feels like the semantics of the message here doesn't quite match the code logic, i.e. The _lookup_ succeeded right after the call to `Info.objects.need_subscribed()`. Whereas here, it feels more accurate to say something like `"Successfully made subscription calls for X users"`

File: openedx/stanford/djangoapps/auth_lagunita/tests/utils.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+    def json(self):
+        """
+        Return content as json-decoded data
+        """
+        data = json_loads(self.content)
+        return data
+
+
+class MockedUser(object):
```
Comment (dcadams) [inline openedx/stanford/djangoapps/auth_lagunita/tests/utils.py]: Why aren't you using UserFactory?

File: openedx/stanford/djangoapps/auth_lagunita/tests/test_form.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
+# pylint: disable=missing-docstring
+from django.test import TestCase
+
+from ..forms import InfoForm
+
+
+class InfoFormTest(TestCase):
+
+    def test_basic(self):
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/tests/test_form.py]: I think it's preferred to have form tests like in `openedx/stanford/djangoapps/register_cme/tests/test_form.py` one passing valid field data and one invalid

File: openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+            self.assertIsNone(lead)
+
+    def test_existing(self):
+        response = MockedResponse(200, json_dumps(DATA['success']))
+        with patch('requests.get', return_value=response):
+            lead = self.api.lookup_lead(self.user.email)
+            self.assertIsNotNone(lead)
+
+    def test_retry_auth_token(self):
+        pass
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py]: Sorry, what's this for?

File: openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+
+    def setUp(self):
+        super(MarketoApiSubscribeTest, self).setUp()
+        self.api = MarketoApi(
+            'localhost',
+            access_token='foobar',
+        )
+        self.api.refresh_access_token = Mock(return_value=self.api.access_token)
+        self.user = MockedUser()
+        self.list_id = 666
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py]: could we add a test for invalid list_id?

File: openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py (hunk 16 above)
Comment (stvstnfrd) [reply openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py]: This was a placeholder;
I have now added a relevant test to ensure that:
- if a request is attempted with an expired/invalid auth_token,
  - it will refresh the token and try the original request one more time

This is to handle the case where the command is long-running and the auth token expires.

File: openedx/stanford/djangoapps/auth_lagunita/api.py (hunk 8 above)
Comment (kluo) [reply openedx/stanford/djangoapps/auth_lagunita/api.py]: I'm speaking to when this code path is reached via `subscribe` from https://github.com/Stanford-Online/edx-platform/pull/903/files#diff-127728bba4e9f71cfc07cffbe6b4ae43R109
and it receives an unrecognized `list_id`

File: openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py [hunk 18]
```diff
@@ -98,20 +98,29 @@ def handle(self, *args, **kwargs):
...
                     info.save()
+                    subscribed += 1
+                else:
+                    failed += 1
             time.sleep(0.5)
         if len(need_subscribed):
             log.info('User lookup succeeded')
         else:
             log.info('User lookup empty')
+        self.stdout.write('Subscription Summary:')
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py]: I think these all need `\n` for newline

File: openedx/stanford/djangoapps/auth_lagunita/api.py (hunk 8 above)
Comment (kluo) [reply openedx/stanford/djangoapps/auth_lagunita/api.py]: Steven tested and confirmed the API returns data with `success:False` and `errors` list for invalid `list_id` errors, so this is fine.

File: openedx/stanford/djangoapps/auth_lagunita/tests/test_form.py (hunk 15 above)
Comment (kluo) [reply openedx/stanford/djangoapps/auth_lagunita/tests/test_form.py]: That's true I guess but then it seems the only necessity to even have tests for Django forms is if they have more complex data validation. Do we simply need this trivial case to satisfy test coverage requirements?

File: openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+            else:
+                subscription = api.subscribe(user, self.list_id)
+                if subscription:
+                    info.submitted_marketing_optin = True
+                    info.save()
+                    subscribed += 1
+                else:
+                    failed += 1
+            time.sleep(0.5)
+        log.info('User lookup completed')
```
Comment (stvstnfrd) [inline openedx/stanford/djangoapps/auth_lagunita/management/commands/opt_in_to_marketing.py]: @kluo I also got rid of `succeeded`;
yeah, this makes more sense :)

File: openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+    'malformed': MockedResponse(HTTP_CODE_SUCCESS, '{"foobar"'),
+    'empty': MockedResponse(HTTP_CODE_SUCCESS, ''),
+    'http_error': MockedResponse(404, ''),
+})
+
+
+class BaseTest(TestCase):
+
+    def setUp(self):
+        super(BaseTest, self).setUp()
```
Comment (kluo) [inline openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py]: I think this call is superfluous?

Comment (stvstnfrd) [reply openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py]: pylint complains without it

**Comment by dcadams** [comment]
:+1: with the additional line for pep8.

File: openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py (hunk 20 above)
Comment (kluo) [reply openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py]: Hm, Django doc examples don't make this call, maybe it's a false positive? Not a big deal tho.

Comment (stvstnfrd) [reply openedx/stanford/djangoapps/auth_lagunita/tests/test_api.py]: Yeah, it could also be configuration difference? 🤷‍♂

**Comment by kluo** [comment]
+1 to squash and merge if tests are passing
